core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/ProgressiveUtils.scala

```diff
@@ -0,0 +1,38 @@
+package edu.uci.ics.texera.workflow.common
+
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, AttributeType}
+
+object ProgressiveUtils {
+
+  // boolean attribute to indicate insertion / retraction
+  // true  indicates insertion  (+)
+  // false indicates retraction (-)
+  val insertRetractFlagAttr = new Attribute("__internal_is_insertion", AttributeType.BOOLEAN)
+
+  def addInsertionFlag(tuple: Tuple): Tuple = {
+    assert(!tuple.getSchema.containsAttribute(insertRetractFlagAttr.getName))
+    Tuple.newBuilder.add(insertRetractFlagAttr, true).add(tuple).build
```

Comment:
I think adding the flag as an attribute is not a good idea. Such internal attributes should be differentiated from data fields. I think we would have more internal attributes for tuples in the future, how about we put them in something like `TupleMeta` that associated with each Tuple? just like `header for `payloads`